pkg/library/status/check.go

```diff
@@ -51,12 +51,34 @@ var unrecoverablePodEventReasons = map[string]int32{
 
 var unrecoverableDeploymentConditionReasons = []string{
 	"FailedCreate",
+	"ProgressDeadlineExceeded",
 }
 
-func CheckDeploymentStatus(deployment *appsv1.Deployment) (ready bool) {
+func CheckDeploymentStatus(deployment *appsv1.Deployment, workspace *common.DevWorkspaceWithConfig) (ready bool) {
+	if workspace.Config.Workspace.DeploymentStrategy == appsv1.RollingUpdateDeploymentStrategyType {
+		return checkRollingUpdateDeploymentStatus(deployment)
+	}
 	return deployment.Status.ReadyReplicas > 0
 }
 
+// Based on the implementation of the `kubectl rollout status deployment` command,
+// see: https://github.com/kubernetes/kubectl/blob/master/pkg/polymorphichelpers/rollout_status.go#L75
+func checkRollingUpdateDeploymentStatus(deployment *appsv1.Deployment) (ready bool) {
+	if deployment.Generation <= deployment.Status.ObservedGeneration {
+		if deployment.Spec.Replicas != nil && deployment.Status.UpdatedReplicas < *deployment.Spec.Replicas {
+			return false
+		}
+		if deployment.Status.Replicas > deployment.Status.UpdatedReplicas {
+			return false
+		}
+		if deployment.Status.AvailableReplicas < deployment.Status.UpdatedReplicas {
+			return false
+		}
+		return true
+	}
+	return false
+}
```

Comment:
I believe this check can be simplified, as the kubectl implementation is designed to report specific information about the state of the rollout. Is it sufficient to check `updatedReplicas == availableReplicas == replicas == 1`?

```suggestion
func checkRollingUpdateDeploymentStatus(deployment *appsv1.Deployment) (ready bool) {
	if deployment.Generation > deployment.Status.ObservedGeneration {
		// Current deployment spec not observed by cluster
		return false
	}
	if deployment.Status.UpdatedReplicas < 1 {
		// Replica has not been updated
		return false
	}
	if deployment.Status.AvailableReplicas < 1 {
		// Updated replica is not available yet
		return false
	}
	if deployment.Status.Replicas != 1 {
		// Old replica is still terminating
		return false // Or true -- the new replica is indeed running and can be accessed
		             // why wait for the old one to terminate before letting the user see
								 // their workspace
	}
	return true
}
```

In particular, we don't really care if the deployment is completed from the cluster's perspective. All we care about is that the updated replica is available (we don't need to wait for termination of the old one necessarily, it _should_ be caught by events if it times out.

Reply:
+1 That's a lot better, thanks for the suggestion :)